ext/AbstractFFTsEnzymeCoreExt.jl

```diff
@@ -0,0 +1,58 @@
+module AbstractFFTsEnzymeCoreExt
+
+using AbstractFFTs
+using AbstractFFTs.LinearAlgebra
+using EnzymeCore
+using EnzymeCore.EnzymeRules
+
+######################
+# Forward-mode rules #
+######################
+
+const DuplicatedOrBatchDuplicated{T} = Union{Duplicated{T},BatchDuplicated{T}}
+
+# since FFTs are linear, implement all forward-model rules generically at a low-level
+
+function EnzymeRules.forward(
+    func::Const{typeof(mul!)},
+    RT::Type{<:Const},
+    y::DuplicatedOrBatchDuplicated{<:StridedArray{T}},
+    p::Const{<:AbstractFFTs.Plan{T}},
+    x::DuplicatedOrBatchDuplicated{<:StridedArray{T}},
+) where {T}
```

Comment:
I wish the type `T` can be restricted to a finite set, e.g. BLAS number types, otherwise, it may produce incorrect gradients for user defined extensions.

Reply:
The pushforward of a linear operator is always itself. And so far as I know, every definition of an FFT is a linear operator. So I can see no reasons why this rule should be problematic for forward-mode.

Reply:
For example, I may want to extended FFT with tropical numbers, which is not a real number. It is linear, but does not have an inverse, you can not assume I do not want to do that. Then your rule would give me incorrect gradients without throwing an error. I have seen too many incorrect gradients in previous AD frameworks such as Zygote when handling complex numbers.

I think there is no reason to claim a rule would work on any non-concrete type. Defining fft rules on BLAS types would be good enough to cover most using cases. For those non-BLAS type, honestly we can not make any assumption for them.

I agree it is good to have a generic backward routine there, but please constraint the interfaces to concrete types when porting it to an AD engine. It should not be so difficult for users to extend the list of supported types in the future.

Reply:
> I may want to extended FFT with tropical numbers

Is this really an FFT per se?  I would consider a [DFT generalized to some other ring](https://en.wikipedia.org/wiki/Discrete_Fourier_transform_over_a_ring) to be a different transform.

Reply:
> > I may want to extended FFT with tropical numbers
> 
> Is this really an FFT per se? I would consider a [DFT generalized to some other ring](https://en.wikipedia.org/wiki/Discrete_Fourier_transform_over_a_ring) to be a different transform.

Since Julia does not have a good trait system, I think it is in general impossible to restrict users to input what the functions are designed for. This is what I meant there lacks provable correctness.

It has been a big issue that none of the Julia libraries (except Enzyme) can provide reliable gradients. They claim too much on untested using cases, like complex numbers and tropical numbers. There has been a belief that "it is cool if it works in cases that it is not expected to work". But no, untested rules are not reliable, they can break on any future change even it works now.

Reply:
By that argument, no AD rules should be defined here anyways, since downstream a user could define a custom Plan that doesn't do any kind of FFT at all. Then even with BLAS number types and strides arrays, any rule we write here would be wrong.

The counterargument is that if a user adds a method of a function whose properties are well-documented, other code should be able to assume and depend on those properties when calling the method for arbitrary inputs.

Taken to its logical conclusion, wouldn't your principle require that rules are never defined for abstract types, and further, that the type of every argument is concrete and known to the rule implementer?

Reply:
> wouldn't your principle require that rules are never defined for abstract types, and further, that the type of every argument is concrete and known to the rule implementer?

A big YES. I do not think many people need the backward rules for non-BLAS types. You may want to support e.g. double float that defined in `DoubleFloat.jl`. I would argue in these using cases, users can port the generic rule to the AD framework with little effort. The rule can be generic, but when porting it to the AD framework, it should be concrete.

It is a bit sad that we have to decide between support more data types and ensure the correctness. I really wish there can be a trait system that user can tell the compiler "this element type is a field", then users can use the rule with more confidence. Facts obvious to you, like "fft should work on field rather than other rings" may not be obvious to others.

> The counterargument is that if a user adds a method of a function whose properties are well-documented, other code should be able to assume and depend on those properties when calling the method for arbitrary inputs.

To differentiate a long code, I will let the code fly and see where it falls. I will add new rules to the AD engine to keep it flying. It is not a problem for me if a rule does not exist. So when using a new element type, like complex number, symbolic type, finite field algebra or the Tropical number type as mentioned above, I will probably not check whether the property of each function is as documented.

> Then even with BLAS number types and strides arrays, any rule we write here would be wrong.

A warning will be thrown when overloading an existing function. Also, pirating is not difficult to avoid.

Reply:
In any case, if we have ChainRules I think we should have the corresponding EnzymeRules.

If users make the questionable choice of overriding `fft` to compute an unrelated function, then it is up to them to override the EnzymeRules/ChainRules as well.